### PR TITLE
[recovery] introduce adaptive recover method

### PR DESCRIPTION
**Pull request by yxieca**
### Description of PR
Summary:
- For interfaces down failure, try to bring up the interfaces.
- For service down failure, reload config to bring them back up.
- For all other failures, reboot to recover.
- Make adaptive the default recover method.
- Restore by reboot if database service is not running. However this change doesn't really take effect because when database is down, failure happens way earlier.

Signed-off-by: Ying Xie <ying.xie@microsoft.com>

### Type of change
- [ ] Bug fix
- [x] Testbed and Framework(new/improvement)
- [ ] Test case(new/improvement)

#### How did you do it?
Add adaptive recover method.

#### How did you verify/test it?
Left a fanout switch port disabled and run any test case. Sanity check recovered the system and proceeded with the test case.
Disable a dut port and run any test case. Sanity check recovered the system and proceeded with the test case.
Disabled a service (pmon), sanity check recovered by config reload and proceeded.

### CODE REVIEW

File: tests/common/plugins/sanity_check/recover.py [hunk 1]
```diff
@@ -26,13 +27,47 @@ def reboot_dut(dut, localhost, cmd):
...
-    wait(30, msg="Wait 30 seconds for system to be stable.")
+    wait(120, msg="Wait 120 seconds for system to be stable.")
 
 
-def recover(dut, localhost, recover_method):
+def __recover_interfaces(dut, fanouthosts, result):
+    for port in result['down_ports']:
+        logging.debug("Restoring port {}".format(port))
+        fanout, fanout_port = fanout_switch_port_lookup(fanouthosts, port)
+        if fanout and fanout_port`:
```
Comment (tahmed-dev) [inline tests/common/plugins/sanity_check/recover.py]: I saw it in the prev diff and while commenting, I got a message that it is outdate. I believe there an extra '`'.

Comment (yxieca) [reply tests/common/plugins/sanity_check/recover.py]: Yeah, I fixed it. Thanks :-)

File: tests/common/plugins/sanity_check/recover.py [hunk 2]
```diff
@@ -26,13 +27,54 @@ def reboot_dut(dut, localhost, cmd):
...
 
-def recover(dut, localhost, recover_method):
+def __recover_interfaces(dut, fanouthosts, result):
+    for port in result['down_ports']:
+        logging.info("Restoring port {}".format(port))
+        fanout, fanout_port = fanout_switch_port_lookup(fanouthosts, port)
+        if fanout and fanout_port:
+            fanout.no_shutdown(fanout_port)
+        dut.no_shutdown(port)
+    wait(30, msg="Wait 30 seconds for interface(s) to restore.")
```
Comment (neethajohn) [inline tests/common/plugins/sanity_check/recover.py:40]: How do we ensure that this fixed the issue and we don't have to go in for a config reload or reboot?

File: tests/common/devices.py [hunk 3]
```diff
@@ -350,6 +350,24 @@ def get_image_info(self):
     def get_asic_type(self):
         return self.facts["asic_type"]
 
+    def shutdown(self, portname):
+        """
+            Shutdown interface (parity function as EosHost)
```
Comment (daall) [inline tests/common/devices.py:355]: I think I get what you're describing here but "parity function as EosHost" is not very clear to me.

File: tests/common/plugins/sanity_check/recover.py [hunk 4]
```diff
@@ -26,13 +27,58 @@ def reboot_dut(dut, localhost, cmd):
...
+    method = constants.RECOVER_METHODS[recover_method]
+    if method["adaptive"]:
+        adaptive_recover(dut, localhost, fanouthosts, check_results)
+    elif method["reboot"]:
         reboot_dut(dut, localhost, constants.RECOVER_METHODS[recover_method]["cmd"])
     else:
-        dut.command(constants.RECOVER_METHODS[recover_method]["cmd"])
-        wait(30, msg="Wait 30 seconds for system to be stable.")
+        dut.command(method["cmd"])
+        wait(60, msg="Wait 60 seconds for system to be stable.")
```
Comment (daall) [inline tests/common/plugins/sanity_check/recover.py:84]: Should this be refactored into a helper like the reboot_dut method now that this "run and wait" bit is being used in multiple places? I think at the least this wait period should be made a constant (e.g. CMD_RECOVERY_WAIT_PERIOD = 60).

File: tests/common/plugins/sanity_check/recover.py [hunk 5]
```diff
@@ -26,13 +27,58 @@ def reboot_dut(dut, localhost, cmd):
         assert False, "Failed to reboot the DUT"
 
     localhost.wait_for(host=dut.hostname, port=22, state="started", delay=10, timeout=300)
-    wait(30, msg="Wait 30 seconds for system to be stable.")
+    wait(120, msg="Wait 120 seconds for system to be stable.")
```
Comment (daall) [inline tests/common/plugins/sanity_check/recover.py:30]: Can you make this a constant (e.g. `REBOOT_RECOVERY_WAIT_PERIOD`)?

File: tests/common/plugins/sanity_check/recover.py (hunk 2 above)
Comment (yxieca) [reply tests/common/plugins/sanity_check/recover.py:40]: Great question!

I believe that if we cannot bring it back up with enabling these 2 ports. Then we won't be able to recover it with config reload either unless something else is wrong. But if something else is wrong, then we should have other sanity check failures that would trigger either config reload or reboot in this adaptive sequence.

File: tests/common/devices.py (hunk 3 above)
Comment (yxieca) [reply tests/common/devices.py:355]: EosHost has shutdown and no_shutdown method to set interface state. To have function with same name would help code unity.

How do you suggest to word this comment?

File: tests/common/plugins/sanity_check/recover.py (hunk 5 above)
Comment (yxieca) [reply tests/common/plugins/sanity_check/recover.py:30]: I further refactored it. I have one more idea, to actually put the function into the method dictionary. But i think I should leave that for another PR.

File: tests/common/plugins/sanity_check/recover.py (hunk 4 above)
Comment (yxieca) [reply tests/common/plugins/sanity_check/recover.py:84]: Made some change. What do you think?

Comment (daall) [reply tests/common/plugins/sanity_check/recover.py]: Looks good to me! :)

**Pull request merged**